pym/bob/layers.py

```diff
@@ -0,0 +1,193 @@
+from .utils import INVALID_CHAR_TRANS
+from .tty import DEBUG, EXECUTED, INFO, NORMAL, IMPORTANT, SKIPPED, WARNING, log
+
+class LayerStepSpec:
+    def __init__(self, path):
+        self.__path = path
+
+    @property
+    def workspaceWorkspacePath(self):
+        return self.__path
+
+    @property
+    def env(self):
+        return {}
+
+    @property
+    def envWhiteList(self):
+        return []
+
+class Layer:
+    def __init__(self, name, root, recipes, yamlCache, scm=None):
+        self.__name = name
+        self.__root = root
+        self.__recipes = recipes
+        self.__yamlCache = yamlCache
+        self.__subLayers = []
+        self.__scm = scm
+        self.__layerDir = os.path.join(self.__root, "layers", self.__name) if len(self.__name) else self.__root
+
+    async def __checkoutTask(self, verbose):
+        if self.__scm is None:
+            return
+        dir = self.__scm.getProperties(False).get("dir")
+        layerSrcPath = os.path.join(self.__root, "layers",
+                                    self.__name)
+        if dir != '.':
+            layerSrcPath = os.path.join(layerSrcPath, dir)
+        invoker = Invoker(spec=LayerStepSpec(layerSrcPath),
+                          preserveEnv=True,
+                          noLogFiles = True,
+                          showStdOut = verbose > INFO,
+                          showStdErr = verbose > INFO,
+                          trace = verbose >= DEBUG,
+                          redirect=False, executor=None)
+        newState = {}
+        newState["digest"] = self.__scm.asDigestScript(),
+        newState["prop"] = {k:v for k,v in self.__scm.getProperties(False).items() if v is not None}
+
+        oldState = BobState().getLayerState(layerSrcPath)
+
+        created = False
+        if not os.path.isdir(layerSrcPath):
+            os.makedirs(layerSrcPath)
+            created = True
+
+        if not created \
+           and self.__scm.isDeterministic() \
+           and oldState is not None \
+           and oldState["digest"] == newState["digest"]:
+            log("CHECKOUT: Layer " +
+                "'{}' skipped (up to date)".format(self.getName()), SKIPPED, INFO)
+            return
+
+        if not created and oldState is not None and \
+            newState["digest"] != oldState["digest"] and \
+            self.__scm.canSwitch(Scm(oldState["prop"],
+                                 self.__recipes.getRootEnv(),
+                                 overrides=[],
+                                 recipeSet=self.__recipes)):
+                ret = await invoker.executeScmSwitch(self.__scm, oldState["prop"])
+                log("SWITCH: Layer '{}' .. ok".format(self.getName()), EXECUTED, INFO)
+
+                if ret == 0:
+                    BobState().setLayerState(layerSrcPath, newState)
+                    return
+        ret = os.path.exists(layerSrcPath)
+
+        if os.path.exists(layerSrcPath) and not created:
```

Comment:
This looks a bit dangerous. IIUC this might move away an existing directory that was placed there before the checkout. In `_cookCheckoutStep` we make sure that we only touch directories that we know of and don't obstruct any unknown directories. We should do the same here unless we have a good reason to deviate.